### PR TITLE
PLIC and CLINT fixes for heterogeneous harts/discontiguous hart IDs

### PR DESCRIPTION
**Pull request by aswaterman**
I went through the PLIC and CLINT to support discontiguous hart IDs.  It turns out the PLIC already supports them, but it wasn't handling the case of some harts having S-mode and some not.

Also, both the PLIC and CLINT were broken on big-endian hosts.  Especially for the CLINT, the fix for that dovetailed with the discontiguous hart ID fix.

### CODE REVIEW

File: riscv/plic.cc [hunk 1]
```diff
@@ -66,21 +67,28 @@
...
-  for (size_t i = 0; i < contexts.size(); i++) {
-    plic_context_t* c = &contexts[i];
-    memset(c, 0, sizeof(*c));
-    c->proc = procs[i / contexts_per_hart];
-    c->mmode = (i % contexts_per_hart == 0);
+  // PLIC contexts are contiguous in memory even if harts are discontiguous.
+  for (const auto& [hart_id, hart] : sim->get_harts()) {
+    contexts.push_back(plic_context_t());
+    plic_context_t* m_context = &contexts.back();
+    memset(m_context, 0, sizeof(*m_context));
```
Comment (scottj97) [inline riscv/plic.cc]: Yuck. Why not let `plic_context_t`'s constructor initialize all the fields? Including `proc` and `mmode`

File: riscv/plic.cc [hunk 2]
```diff
@@ -340,17 +340,15 @@ bool plic_t::load(reg_t addr, size_t len, uint8_t* bytes)
     }
   }
 
-  if (ret) {
```
Comment (scottj97) [inline riscv/plic.cc:343]: The removal of the conditional here should be its own commit, so in case it happens to cause some problem, `git bisect` can pinpoint it.

File: riscv/plic.cc [hunk 3]
```diff
@@ -340,17 +340,15 @@ bool plic_t::load(reg_t addr, size_t len, uint8_t* bytes)
...
-    memcpy(bytes, (uint8_t *)&val, len);
-  }
+  read_little_endian_reg(val, addr, len, bytes);
```
Comment (scottj97) [inline riscv/plic.cc:332]: 09db15baf422b0af769effe3ad5f09b722f619cf contains only dead code, which impairs `git bisect`; I'd squash it with this commit.

**Review by scottj97 — CHANGES_REQUESTED** [review]

File: riscv/clint.cc [hunk 4]
```diff
@@ -29,16 +30,27 @@ clint_t::clint_t(std::vector<processor_t*>& procs, uint64_t freq_hz, bool real_t
...
-    memcpy(bytes, (uint8_t*)&mtime + addr - MTIME_BASE, len);
+
+  if (addr >= MSIP_BASE && addr < MTIMECMP_BASE) {
+    if (len == 8) {
+      // Implement double-word loads as a pair of word loads
+      return load(addr, 4, bytes) && load(addr + 4, 4, bytes + 4);
+    }
+
+    const auto hart_id = (addr - MSIP_BASE) / sizeof(msip_t);
+    const msip_t res = sim->get_harts().count(hart_id) ? sim->get_harts().at(hart_id)->state.mip->read() & MIP_MSIP : 0;
```
Comment (scottj97) [inline riscv/clint.cc]: This causes MSIP to read out as 8 instead of 1

File: riscv/plic.cc (hunk 3 above)
Comment (aswaterman) [reply riscv/plic.cc:332]: Ack.

File: riscv/clint.cc (hunk 4 above)
Comment (aswaterman) [reply riscv/clint.cc]: Oops. Thanks for catching that.

File: riscv/plic.cc (hunk 1 above)
Comment (aswaterman) [reply riscv/plic.cc]: Yes, that’s better still.

File: riscv/plic.cc (hunk 2 above)
Comment (aswaterman) [reply riscv/plic.cc:343]: Ack.

File: riscv/plic.cc (hunk 3 above)
Comment (aswaterman) [reply riscv/plic.cc:332]: For my edification, why do commits like this impair bisect?  Is it merely that a longer history makes it take longer to bisect, or is there something more subtle that I'm missing?

Comment (scottj97) [reply riscv/plic.cc:332]: If there was a bug in the dead code, `git bisect` would point to the following commit (where the dead code was first executed) instead of the commit where the bug was actually introduced.

Edit: it's not just for git bisect, though; I want every commit to be perfect, and dead code is not perfect. But perhaps that's overly pedantic.

Comment (aswaterman) [reply riscv/plic.cc:332]: Thanks for the explanation.  I think we should run with your preferred approach as a guideline, not a hard-and-fast rule.  Sometimes, the dead code is substantial enough that separating it into its own commit is justifiable on the basis of reducing cognitive load.  (This isn't such a case.)

File: riscv/devices.h [hunk 5]
```diff
@@ -79,6 +79,11 @@ class clint_t : public abstract_device_t {
 #define PLIC_MAX_DEVICES 1024
 
 struct plic_context_t {
+  plic_context_t(processor_t* proc, bool mmode)
+    : proc(proc), mmode(mmode), priority_threshold(0), enable{}, pending{},
```
Comment (scottj97) [inline riscv/devices.h:85]: Is there any significance to the mix of parentheses and curly braces?

Comment (scottj97) [reply riscv/devices.h:85]: I think it would be preferable to use in-class member initializers for everything but `proc` and `mmode`. E.g:
```
uint8_t priority_threshold{};
uint32_t enable[PLIC_MAX_DEVICES/32]{};
```

**Review by scottj97 — APPROVED** [review]

File: riscv/clint.cc [hunk 6]
```diff
@@ -29,16 +30,27 @@ clint_t::clint_t(std::vector<processor_t*>& procs, uint64_t freq_hz, bool real_t
...
+
+  if (addr >= MSIP_BASE && addr < MTIMECMP_BASE) {
+    if (len == 8) {
+      // Implement double-word loads as a pair of word loads
+      return load(addr, 4, bytes) && load(addr + 4, 4, bytes + 4);
+    }
+
+    const auto hart_id = (addr - MSIP_BASE) / sizeof(msip_t);
+    const msip_t res = sim->get_harts().count(hart_id) && (sim->get_harts().at(hart_id)->state.mip->read() & MIP_MSIP);
+    read_little_endian_reg(res, addr, len, bytes);
```
Comment (scottj97) [inline riscv/clint.cc:46]: If there's another device (or memory) just before MSIP_BASE, a misaligned `LD` to MSIP_BASE-2 could come here with len==6, resulting in abort:
```
spike: riscv-isa-sim/riscv/devices.h:176: void read_little_endian_reg(T, reg_t, size_t, uint8_t*) [with T = unsigned int; reg_t = long unsigned int; size_t = long unsigned int; uint8_t = unsigned char]: Assertion `len <= sizeof(T)' failed.
```

File: riscv/devices.h (hunk 5 above)
Comment (aswaterman) [reply riscv/devices.h:85]: Agreed; I'm mentally stuck in C++03.

File: riscv/clint.cc (hunk 6 above)
Comment (aswaterman) [reply riscv/clint.cc:46]: It sure does.  I had been thinking that misaligned accesses cannot make it here, and I would be surprised if this is the only I/O device with a bug around that assumption.

File: riscv/devices.h (hunk 5 above)
Comment (aswaterman) [reply riscv/devices.h:85]: > Is there any significance to the mix of parentheses and curly braces?

I was misremembering that `{}` was syntax for array initialization only.
